### PR TITLE
Create a Readme in `lib`

### PR DESCRIPTION
**Pull request by infinisil**
###### Description of changes
Long overdue. Motivated by trying to document the Nixpkgs structure in https://github.com/NixOS/nixpkgs/pull/237439, which could benefit from Readme's going into more detail.

### CODE REVIEW

File: lib/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  Example: `lib.lists` is a sub-library containing list-related functionality such as `lib.lists.take` and `lib.lists.imap0`.
+  These are defined in the file [`lists.nix`](lists.nix).
+
+- Aliases:
+  Attributes that point to an attribute of the same name in some sub-library.
+
+  Example: `lib.take` is an alias for `lib.lists.take`.
+
+Most files this directory are definitions of sub-libraries, but there are a few others:
```
Comment (proofconstruction) [inline lib/README.md:21]:
```suggestion
Most files in this directory are definitions of sub-libraries, but there are a few others:
```

File: lib/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+- [`minver.nix`](minver.nix): A string of the minimum version of Nix that is required to evaluate Nixpkgs.
+- [`tests`](tests): Tests, see [Running tests](#running-tests)
+  - [`release.nix`](tests/release.nix): A derivation aggregating all tests
+  - [`misc.nix`](tests/misc.nix): Evaluation unit tests for most sub-libraries
+  - `*.sh`: Bash script that run tests for specific sub-libraries
+  - Rest: Support files for the tests
```
Comment (proofconstruction) [inline lib/README.md:27]:
```suggestion
  - All other files in this directory exist to support the tests
```

This doesn't fit the format of the previous lines, which is actually ideal in this case.

File: lib/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  Attributes that point to an attribute of the same name in some sub-library.
+
+  Example: `lib.take` is an alias for `lib.lists.take`.
+
+Most files this directory are definitions of sub-libraries, but there are a few others:
+- [`minver.nix`](minver.nix): A string of the minimum version of Nix that is required to evaluate Nixpkgs.
+- [`tests`](tests): Tests, see [Running tests](#running-tests)
+  - [`release.nix`](tests/release.nix): A derivation aggregating all tests
+  - [`misc.nix`](tests/misc.nix): Evaluation unit tests for most sub-libraries
+  - `*.sh`: Bash script that run tests for specific sub-libraries
```
Comment (proofconstruction) [inline lib/README.md:26]:
```suggestion
  - `*.sh`: Bash scripts that run tests for specific sub-libraries
```

File: lib/README.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  - Rest: Support files for the tests
+- [`systems`](systems): The `lib.systems` sub-library, structured into a directory instead of a file due to its complexity
+- [`path`](path): The `lib.path` sub-library including design document and tests
```
Comment (proofconstruction) [inline lib/README.md:29]:
```suggestion
- [`path`](path): The `lib.path` sub-library, which includes tests as well as a document describing the design goals of `lib.path`
```

File: lib/README.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+- Rest: Sub-libraries
```
Comment (proofconstruction) [inline lib/README.md:30]:
```suggestion
- All other files in this directory are sub-libraries
```

File: lib/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+Reference documentation for library functions is written above each function as a multi-line comment.
+These comments are processed using [nixdoc](https://github.com/nix-community/nixdoc) and [rendered in the Nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#chap-functions).
+The comment format is described in the nixdoc README [here](https://github.com/nix-community/nixdoc#comment-format).
+
+See the [chapter on contributing to the manual](https://nixos.org/manual/nixpkgs/#chap-contributing) for how to build the manual.
+
+## Running tests
+
+All library tests can run by building the derivation in [`tests/release.nix`](tests/release.nix):
```
Comment (proofconstruction) [inline lib/README.md:49]:
```suggestion
All library tests can be run by building the derivation in [`tests/release.nix`](tests/release.nix):
```

File: lib/README.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+```bash
+nix-build tests/release.nix
+```
+
+Some commands for quicker iteration over parts are also available:
```
Comment (proofconstruction) [inline lib/README.md:55]:
```suggestion
Some commands for quicker iteration over parts of the test suite are also available:
```

**Review by proofconstruction — COMMENTED** [review]
A sorely-needed addition to the project documentation!

These are just some suggestions to improve phrasing/clarity and fix typos.

**Review by zmitchell — APPROVED** [review]

File: lib/README.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
+# Nixpkgs lib
+
+This folder contains the implementation, documentation and tests for the Nixpkgs `lib` library.
```
Comment (fricklerhandwerk) [inline lib/README.md:3]:
```suggestion
This directory contains the implementation, documentation and tests for the Nixpkgs `lib` library.
```

This is completely minor, but using "directory" consistently would probably help beginners in particular.

File: lib/README.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multiple sub-libraries:
+- [`modules.nix`](modules.nix): `lib.modules` for the core functions and anything not relating to option definitions
+- [`options.nix`](options.nix): `lib.options` for anything relating to option definitions
+- [`types.nix`](types.nix): `lib.types` for module system types
+
+## Reference documentation
+
+Reference documentation for library functions is written above each function as a multi-line comment.
+These comments are processed using [nixdoc](https://github.com/nix-community/nixdoc) and [rendered in the Nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#chap-functions).
+The comment format is described in the nixdoc README [here](https://github.com/nix-community/nixdoc#comment-format).
```
Comment (fricklerhandwerk) [inline lib/README.md:43]:
```suggestion
The nixdoc README describes the [comment format](https://github.com/nix-community/nixdoc#comment-format).
```

File: lib/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+See the [chapter on contributing to the manual](https://nixos.org/manual/nixpkgs/#chap-contributing) for how to build the manual.
```
Comment (fricklerhandwerk) [inline lib/README.md:45]:
```suggestion
See the [chapter on contributing to the Nixpkgs manual](https://nixos.org/manual/nixpkgs/#chap-contributing) for how to build the manual.
```

We have to be explicit until further notice.

**Review by fricklerhandwerk — APPROVED** [review]
Great stuff!

**Comment by infinisil** [comment]
Thanks for the fast reviews @proofconstruction @fricklerhandwerk and @zmitchell, I agree with the suggestions and applied them :)

File: lib/README.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+All library tests can be run by building the derivation in [`tests/release.nix`](tests/release.nix):
+
+```bash
+nix-build tests/release.nix
+```
+
+Some commands for quicker iteration over parts of the test suite are also available:
+
+```bash
+# Run all evaluation unit tests in tests/misc.nix
```
Comment (roberth) [inline lib/README.md:58]:
```suggestion
# Run all evaluation unit tests in tests/misc.nix
# if the resulting list is empty, all tests passed
```

See the file.

File: lib/README.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+# Run the lib.sources tests
+tests/sources.sh
+
+# Run the lib.filesystem tests
+tests/filesystem.sh
+
+# Run the lib.path property tests
+path/tests/prop.sh
+```
```
Comment (roberth) [inline lib/README.md:72]: This should be documented at least in the files themselves, but having a copy here is probably good.

Comment (infinisil) [reply lib/README.md:72]: How about just linking to this README from those files to avoid duplicated information?

Comment (roberth) [reply lib/README.md:72]: I think the loss of convenience due to the indirection outweighs the duplication, because this info doesn't change often.

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
Awesome, thanks!

**Pull request merged**